File: OJ/leet421/leet421.cpp
/*
leet421


*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <utility>
// #include "../utils/LeetCpp.utils.hpp"

using namespace std;

class Solution {
public:
    int findMaximumXOR1(const vector<int>& nums) {
        int ans = 0;
        for (int i = 31; i >= 0; --i) {
            ans <<= 1;
            auto curXor = ans | 1;
            unordered_set<int> S;
            for (const auto& x: nums) S.insert(x >> i);
            for (const auto& x: S) {
                if (S.count(curXor ^ x)) { ans = curXor; break; }
            }
        }
        return ans;
    }
    int findMaximumXOR(const vector<int>& nums) {
        // 80ms, O(N log N)
        // 从最高位开始，逐位向低处检查
        int ans = 0;
        for (int i = 30; i >= 0; --i) {
            ans <<= 1;
            auto curXor = ans | 1; // candidate 
            unordered_set<int> S;
            for (const auto& x: nums) {
                if (S.count(curXor ^ (x >> i))) { ans = curXor; break; }
                S.insert(x >> i);
            }
        }
        return ans;
    }
};


int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << sol.findMaximumXOR({3, 10, 5, 25, 2, 8}) << endl;
    cout << sol.findMaximumXOR({8, 10, 2}) << endl;
    return 0;
}

File: OJ/leet421/leet421.py
# leet421

from typing import List


class Solution:
    def findMaximumXOR1(self, nums: List[int]) -> int:
        # 哈希法，
        # 要点 x ^ a = b  =>  x ^ b = a,  a ^ b = x;
        L = max(len(bin(x)) for x in nums)
        ans = 0
        for i in range(L)[::-1]:
            ans <<= 1
            curXor = ans | 1;
            test = set(x >> i for x in nums)
            ans = ans | any(curXor ^ x in test for x in test)
        return ans

    def findMaximumXOR(self, nums: List[int]) -> int:
        # 字典树法
        # 对每个数在 Trie 中寻找最大的互补对
        L = max(len(bin(x)) for x in nums)
        ans = 0
        for i in range(L)[::-1]:
            ans <<= 1
            curXor = ans | 1;
            test = set(x >> i for x in nums)
            ans = ans | any(curXor ^ x in test for x in test)
        return ans







print(Solution().findMaximumXOR([3, 10, 5, 25, 2, 8]))
print(Solution().findMaximumXOR([8, 10, 2]))


